File: algorithms/95.cpp
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution
{
private:
    void dfs(int start, int end, vector<TreeNode *> &result, vector< vector< vector<TreeNode *> > > &dp)
    {
        if ( start > end )
        {
            result.push_back(NULL);
            return;
        }
        if ( !dp[start-1][end-1].empty() )
        {
            result = dp[start-1][end-1];
            return;
        }
        for ( int i = start; i <= end; ++i )
        {
            vector<TreeNode *> left, right;
            dfs(start, i-1, left, dp);
            dfs(i+1, end, right, dp);
            for ( auto left_node = left.begin(); left_node != left.end(); ++left_node )
            {
                for ( auto right_node = right.begin(); right_node != right.end(); ++right_node )
                {
                    TreeNode *node = new TreeNode(i);
                    node->left = *left_node;
                    node->right = *right_node;
                    result.push_back(node);
                }
            }
        }
        dp[start-1][end-1] = result;
        return;
    }
public:
    vector<TreeNode*> generateTrees(int n)
    {
        vector<TreeNode *> result;
        vector< vector< vector<TreeNode *> > > dp(n, vector< vector<TreeNode *> >(n));
        dfs(1, n, result, dp);
        return result;
    }
};

int main()
{
    return 0;
}

File: algorithms/rust/95.rs
use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

impl Solution {
    #[allow(unused)]
    pub fn generate_trees(n: i32) -> Vec<Option<Rc<RefCell<TreeNode>>>> {
        Self::build(1, n)
    }

    fn build(left: i32, right: i32) -> Vec<Option<Rc<RefCell<TreeNode>>>> {
        if left > right {
            return vec![None];
        }
        let mut rslt = vec![];
        for mid in left..=right {
            for left_tree in Self::build(left, mid - 1) {
                for right_tree in Self::build(mid + 1, right) {
                    let mut root = TreeNode::new(mid);
                    root.left = left_tree.to_owned();
                    root.right = right_tree.to_owned();
                    rslt.push(Some(Rc::new(RefCell::new(root))));
                }
            }
        }
        rslt
    }
}

struct Solution;

fn main() {
    println!("Hello, world!");
}
